src/nrnpython/nrnpy_nrn.cpp

```diff
@@ -1493,29 +1490,28 @@ static PyObject* NPySecObj_connect_safe(NPySecObj* self, PyObject* args) {
 static PyObject* NPySecObj_insert(NPySecObj* self, PyObject* args) {
     CHECK_SEC_INVALID(self->sec_);
     char* tname;
-    PyObject *tpyobj, *tpyobj2;
     if (!PyArg_ParseTuple(args, "s", &tname)) {
         PyErr_Clear();
         // if called with an object that has an insert method, use that
+        PyObject* tpyobj;
         if (PyArg_ParseTuple(args, "O", &tpyobj)) {
-            Py_INCREF(tpyobj);
-            Py_INCREF((PyObject*) self);
-            tpyobj2 = PyObject_CallMethod(tpyobj, "insert", "O", (PyObject*) self);
-            Py_DECREF(tpyobj);
-            if (tpyobj2 == NULL) {
-                Py_DECREF((PyObject*) self);
+            auto _tpyobj_tracker = nb::borrow(tpyobj);
+            // tpyobj.insert reqs incref self. Borrow it so if it fails the ref count rolls-back
+            auto self_obj = nb::borrow((PyObject*) self);
+            auto tpyobj2 = nb::steal(PyObject_CallMethod(tpyobj, "insert", "O", self_obj.ptr()));
+            if (!tpyobj2.is_valid()) {
                 PyErr_Clear();
                 PyErr_SetString(
                     PyExc_TypeError,
                     "insert argument must be either a string or an object with an insert method");
-                return NULL;
+                return nullptr;
             }
-            Py_DECREF(tpyobj2);
-            return (PyObject*) self;
+            self_obj.inc_ref();  // We return the object, need INCREF again
+            return self_obj.release().ptr();
```

Comment:
I didn't find a good source, but this info:

Borrowed References:

From the [Python/C API Reference Manual](https://docs.python.org/3/c-api/intro.html#borrowed-references):
A borrowed reference is a reference to an object. The borrower (i.e., the code which receives the reference) does not need to increment the reference count. However, the borrower must not Py_DECREF() the object. Borrowed references are often used when objects are retrieved from other objects.

Reference Counting and Argument Parsing:

From the [Extending and Embedding the Python Interpreter](https://docs.python.org/3/extending/extending.html#reference-counts):
When creating a new object, its reference count is set to 1. When passing an object reference into a function, there is no need to increment the reference count: the function receives ownership of the reference.

PyArg_ParseTuple and Argument Handling:

From the [PyArg_ParseTuple](https://docs.python.org/3/c-api/arg.html#c.PyArg_ParseTuple) documentation:
PyArg_ParseTuple() and its variants do not increment the reference count of objects passed into it; they pass borrowed references to the parameters specified in the format string.

Putting these pieces together, the key points are:

 - When a C function receives a Python object as an argument, it typically receives a borrowed reference.
 - A borrowed reference allows the function to use the object without incrementing the reference count.
 - If the function needs to retain a reference to the object beyond its scope, it should use Py_INCREF to increment the reference count.
 - It is the function's responsibility to manage the reference count correctly to avoid memory leaks or dangling pointers.

Reply:
Okay, that seems sensible. Then why are we bumping the reference count to `+2` by the time we return from this function?

Reply:
Because we have two new references
 - the "inserted" one
 - the returned one

Reply:
Is there a sanitizer for this?

I'm not sure I fully understand the argument:
1. When we "insert" we're calling `PyObject_CallMethod(tpyobj, "insert", "O", self_obj.ptr())`. Therefore, I'd argue that `PyObject_CallMethod` is a C function and therefore receives a borrowed reference. It will lead to calling something that actually does the insert, *that* function would need to bump the reference count of `self_obj`.
2. We return a borrowed reference.

Hence I would have thought that we should leave at `+0`. I could see `+1` if we return a "new reference".

Reply:
We can't return a borrowed reference without INCREF, because it is a new reference going outside the scope. About that I have no doubts and luckily we see that being respected in the rest of the file.

For the "insert" you have a good point. From the original code we have a hint that it is up to us to incref, hence they DECREF in case the insert fails. But this assumption might be wrong and the decref might be there because we would be returning null. I need to investigate too.

Reply:
(I lost a comment, so recreating it here.)

I found the documentation for returning a "new reference". It's here:
https://docs.python.org/3/c-api/structures.html#c.PyCFunction

This justifies `+1` which is what NRN currently does. The DECREF is because of error handling. It bumps the reference early just to be safe (after all we're calling into the Python library and it might clean up the reference form under us). Then on the error path it need to decrement the reference again. Finally, it leaves with `+1` on the thing it returns, i.e. it return a "new reference".

Reply:
We are not returning a new reference, we are returning self. I'm confused why you'r mentioning that

Reply:
Let me do some investigation and then we resume the review with more details :)

Reply:
Ahhhh... ok gotcha
```
NPySecObj_insert_safe( self ) {     /// received self is a borrowed reference (we wont steal it)
    ....
   return X  /// it is expected X to be a new reference - this is almost always the case so that the function can be used from the interpreter and keep the reference
}